### ISSUE TITLE
CMakeLists add OpenBLAS as subdirectory and include directories

### ISSUE THREAD

**Issue by j-c-cook** [xianyi/OpenBLAS#3240]
I am using `OpenBLAS` for my `BLAS` and `LAPACK` routines because I have found it to be exceedingly faster than `netlib`. However, I do not want the userbase of my `C++` library to have to install this dependency themselves. I would like to incorporate `OpenBLAS` into my `third_party/` folder and then call `add_subdirectory(third_party/OpenBLAS)` in CMakeLists.txt. 

 Upon building with CMake this way, I get the following warning from OpenBLAS
```
CMake Warning at third_party/OpenBLAS/CMakeLists.txt:70 (message):
  CMake support is experimental.  It does not yet support all build options
  and may not produce the same Makefiles that OpenBLAS ships with.
```
This message by itself is not worrisome. 

However, typically with `C++` libraries there is a `include/` directory, which can be defined in CMake as `include_directories(include)`. Is there a specific folder I should include? I am unable to call the routines I typically make use of. 

Here's an example of a routine I am using and calling from C++
```
extern "C" void daxpy_(int *n, double *a, double *x, int *incx, double *y, int *incy);
daxpy_(&n, &a, &*x.begin(), &incx, &*y.begin(), &incy);
```
Where `x` and `y` are both of container `std::vector<double>`.

@hpanderson

**Comment by martin-frbg** [comment]
This may be a bit problematic right now as the headers live in different subdirectories of the build - some of the files are generated - and are collected into a common include directory only in the `install` step of the OpenBLAS build. From the interface you quoted, you are probably looking for the file f77blas.h (which starts life in the "generated" subdirectory but  has a dependency on openblas_config.h from the toplevel of the build )

**Comment by martin-frbg** [comment]
That said, I am not sure why you would be "unable to call the routines" with the `extern "C"` declaration you quoted for daxpy (unless your compiler chose to generate symbol names without a trailing underscore)

**Comment by j-c-cook** [comment]
Its because in addition to defining `add_subdirectories(third_party/OpenBLAS)`, I also need to define a `include_directories(path/to/include)`, where `path/to/include` contains header files. 

@martin-frbg I [posted](https://stackoverflow.com/questions/67651045/how-to-define-openblas-as-a-sub-directory-in-cmake) on stack overflow to see if anyone has any ideas. I may have described my problem with better details there.

**Comment by martin-frbg** [comment]
I doubt you will actually need the include path to the OpenBLAS header(s) as long as you declare the BLAS interfaces in your own source - to me this looks more like it cannot find the libopenblas.a (or .so) itself for linking. (I am by no means an expert in cmake, but I do not see how it is expected to figure out the `target_link_libraries` on its own with no BLAS_LIBRARIES given in the non-working example. I suspect something like `target_link_libraries(OBLASCmake third_party/OpenBLAS-0.3.15/build/libopenblas.a` would be needed for that to work, probably with ${PROJECT_SOURCE_DIR} or ${PROJECT_BINARY_DIR} thrown in to fix up the relative path to your third_party)

**Comment by xq114** [comment]
Why not use package managers? Like vcpkg, conan, xrepo, they are good for simplifying installing of external packages

**Comment by martin-frbg** [comment]
Did you get this sorted ? The parallel stackoverflow thread with its morphing example was a bit hard to follow

**Comment by j-c-cook** [comment]
No I haven't got it figured out yet, but I think I will figure something out.
